### DISCUSSION TITLE
Duplicate detection in new save logic

### DISCUSSION THREAD

**Discussion by runningjustin** [babyfish-ct/jimmer#663] [General]
I noticed that with the new save logic (0.8.170) there's a SaveException.NotUnique exception thrown for non-uniqueness. In the previous version (0.8.132) we were using the e.cause to get the PSQLException and parse the error message. This is helpful as it tells us the index name so we can provide a more specific error message. Is it still possible to provide the database driver exception or are you not even getting this in the new version?

**Comment by babyfish-ct** [comment]
For old version, No exception translator

For new version, duplicate id, duplicate key *(If there is unique constraint)* and illegal associated id *(If the foreign key is real, that means there are foreign constraints)* will be automatically translated by Jimmer, that is `SaveException.NotUnique` or `SaveException.IllegalTargetId`, other database exceptions will not be handed

No matter SQL exception is translated for not, you can use [ExceptionTranslator](https://github.com/babyfish-ct/jimmer/blob/main/project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/runtime/ExceptionTranslator.java) to handle it.

**Reply by runningjustin** [reply]
OK, thanks!
